File: apps/Offline-AI-Buddy/shared/BuddyAICore/Sources/BuddyAICore/Domain/QuotaTracker.swift
import Foundation

/// Pure `(state, event) -> state` reducer for the per-profile, per-day
/// chat quota. Free tier = 10 chats/day; each watched ad = +5 chats;
/// pro entitlement bypasses everything.
public enum QuotaTracker {

    public static let freeDailyLimit: Int = 10
    public static let chatsPerAd: Int = 5

    public enum Event: Hashable, Sendable {
        case chatStarted
        case adWatched
        /// Force-roll the day forward (used at midnight / app foreground).
        case rollover(toDay: String)
    }

    public struct Decision: Hashable, Sendable {
        public let allowed: Bool
        public let chatsRemaining: Int
        public let canWatchAd: Bool

        public init(allowed: Bool, chatsRemaining: Int, canWatchAd: Bool) {
            self.allowed = allowed
            self.chatsRemaining = chatsRemaining
            self.canWatchAd = canWatchAd
        }
    }

    /// Apply an event. Pro-unlocked profiles bypass quota checks.
    public static func reduce(
        state: QuotaState,
        event: Event,
        proUnlocked: Bool
    ) -> QuotaState {
        switch event {
        case .chatStarted:
            if proUnlocked { return state }
            var s = state
            s.chatsUsed += 1
            return s
        case .adWatched:
            if proUnlocked { return state }
            var s = state
            s.adUnlocks += 1
            return s
        case .rollover(let day):
            return QuotaState(profileId: state.profileId, day: day, chatsUsed: 0, adUnlocks: 0)
        }
    }

    /// Decide whether the next chat is allowed.
    public static func decide(state: QuotaState, proUnlocked: Bool) -> Decision {
        if proUnlocked {
            return Decision(allowed: true, chatsRemaining: Int.max, canWatchAd: false)
        }
        let allowance = freeDailyLimit + state.adUnlocks * chatsPerAd
        let remaining = max(0, allowance - state.chatsUsed)
        return Decision(allowed: remaining > 0, chatsRemaining: remaining, canWatchAd: true)
    }
}

File: apps/Offline-AI-Buddy/android/core/src/main/java/com/americangroupllc/offlineaibuddy/core/domain/QuotaTracker.kt
package com.americangroupllc.offlineaibuddy.core.domain

import com.americangroupllc.offlineaibuddy.core.models.QuotaState

/**
 * Pure (state, event) -> state reducer. Mirrors `BuddyAICore.QuotaTracker`.
 */
object QuotaTracker {

    const val FREE_DAILY_LIMIT = 10
    const val CHATS_PER_AD = 5

    sealed interface Event {
        data object ChatStarted : Event
        data object AdWatched : Event
        data class Rollover(val toDay: String) : Event
    }

    data class Decision(
        val allowed: Boolean,
        val chatsRemaining: Int,
        val canWatchAd: Boolean,
    )

    fun reduce(state: QuotaState, event: Event, proUnlocked: Boolean): QuotaState =
        when (event) {
            is Event.ChatStarted ->
                if (proUnlocked) state else state.copy(chatsUsed = state.chatsUsed + 1)
            is Event.AdWatched ->
                if (proUnlocked) state else state.copy(adUnlocks = state.adUnlocks + 1)
            is Event.Rollover ->
                QuotaState(profileId = state.profileId, day = event.toDay)
        }

    fun decide(state: QuotaState, proUnlocked: Boolean): Decision {
        if (proUnlocked) return Decision(true, Int.MAX_VALUE, false)
        val allowance = FREE_DAILY_LIMIT + state.adUnlocks * CHATS_PER_AD
        val remaining = (allowance - state.chatsUsed).coerceAtLeast(0)
        return Decision(remaining > 0, remaining, true)
    }
}
